pkg/statedb/reconciler/reconciler_test.go

```diff
@@ -0,0 +1,538 @@
+					return mt, mt
+				},
+			),
+			cell.Provide(func() reconciler.Config[*testObject] {
+				return reconciler.Config[*testObject]{
+					// Don't run the full reconciliation via timer, but rather explicitly so that the full
+					// reconciliation operations don't mix with incremental when not expected.
+					FullReconcilationInterval: time.Hour,
+
+					RetryBackoffMinDuration: time.Millisecond,
+					RetryBackoffMaxDuration: 10 * time.Millisecond,
+					IncrementalBatchSize:    1000,
+					GetObjectStatus:         (*testObject).GetStatus,
+					WithObjectStatus:        (*testObject).WithStatus,
+					StatusIndex:             statusIndex,
+				}
+			}),
+			cell.Provide(reconciler.New[*testObject]),
+
+			cell.Invoke(func(r_ reconciler.Reconciler[*testObject], h cell.Health) {
+				r = r_
+				health = h
+			}),
+		),
+	)
+
+	require.NoError(t, hive.Start(context.TODO()), "Start")
+
+	h := testHelper{
+		t:      t,
+		db:     db,
+		tbl:    testObjects,
+		ops:    mt,
+		r:      r,
+		health: health,
+	}
+
+	numIterations := 10
+
+	t.Run("incremental", func(t *testing.T) {
+		h.t = t
+
+		for i := 0; i < numIterations; i++ {
+			t.Logf("Iteration %d", i)
+
+			// Insert some test objects and check that they're reconciled
+			t.Logf("Inserting test objects 1, 2 & 3")
+			h.insert(ID_1, NonFaulty, reconciler.StatusPending())
+			h.expectOp("update: 1")
+			h.expectStatus(ID_1, reconciler.StatusKindDone, "")
+
+			h.insert(ID_2, NonFaulty, reconciler.StatusPending())
+			h.expectOp("update: 2")
+			h.expectStatus(ID_2, reconciler.StatusKindDone, "")
+
+			h.insert(ID_3, NonFaulty, reconciler.StatusPending())
+			h.expectOp("update: 3")
+			h.expectStatus(ID_3, reconciler.StatusKindDone, "")
+
+			h.expectHealthLevel(cell.StatusOK)
+
+			// Set one to be faulty => object will error
+			t.Log("Setting '1' faulty")
+			h.insert(ID_1, Faulty, reconciler.StatusPending())
+			h.expectOp("update fail: 1")
+			h.expectStatus(ID_1, reconciler.StatusKindError, "update fail")
+			h.expectRetried(ID_1)
+			h.expectHealthLevel(cell.StatusDegraded)
+
+			// Fix the object => object will reconcile again.
+			t.Log("Setting '1' non-faulty")
+			h.insert(ID_1, NonFaulty, reconciler.StatusPending())
+			h.expectOp("update: 1")
+			h.expectStatus(ID_1, reconciler.StatusKindDone, "")
+			h.expectHealthLevel(cell.StatusOK)
+
+			t.Log("Delete 1 & 2")
+			h.markForDelete(ID_1)
+			h.expectOp("delete: 1")
+			h.expectNotFound(ID_1)
+
+			h.markForDelete(ID_2)
+			h.expectOp("delete: 2")
+			h.expectNotFound(ID_2)
+
+			t.Log("Try to delete '3' with faulty ops")
+			h.setTargetFaulty(true)
+			h.markForDelete(ID_3)
+			h.expectOp("delete fail: 3")
+			h.expectStatus(ID_3, reconciler.StatusKindError, "delete fail")
+			h.expectHealthLevel(cell.StatusDegraded)
+
+			t.Log("Delete 3")
+			h.setTargetFaulty(false)
+			h.expectOp("delete: 3")
+			h.expectNotFound(ID_3)
+			h.expectHealthLevel(cell.StatusOK)
+
+		}
+	})
+
+	t.Run("full", func(t *testing.T) {
```

Comment:
This afaict doesn't cover the external deletion case, can we test that we detect out of sync things in that case?

Reply:
... and the out-of-sync metric is broken if full reconciliation of objects fail and are retried. hmm not sure how to cleanly fix this. might need to keep track of whether the item was queued from incremental or full reconciliation.

Reply:
Yeah I guess it depends on the intended semantics of the metric - if we want to say 'how many times we encountered items which are out of sync' then I think it's fine if we count the same multiple times.

Reply:
The expected use-case for the metric is to essentially look whether "outOfSync > 0" within some time span and then investigate further if that's the case. E.g. we'd use this to catch if systemd is messing with us.